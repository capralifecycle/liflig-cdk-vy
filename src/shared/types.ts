/**
 * Shared type definitions for Vy custom resources
 */

/**
 * The different Vy environments
 */
export enum VyEnvironment {
  /**
   * Development environment
   */
  TEST = "test",

  /**
   * Production-like environment
   */
  STAGE = "stage",

  /**
   * Production environment
   */
  PROD = "prod",
}

export interface Scope {
  name: string
  description: string
}

export interface ResourceServer {
  identifier: string
  name: string
  scopes?: Scope[]
}

export interface ResourceServerUpdateRequest {
  identifier: string
  name: string
  scopes?: Scope[]
}

export interface AppClient {
  name: string
  scopes: string[]
  type: "frontend" | "backend"
  callback_urls: string[]
  logout_urls: string[]
  generate_secret?: boolean
  client_id?: string
  client_secret?: string
}

export interface AppClientUpdateRequest {
  name: string
  scopes: string[]
  callback_urls: string[]
  logout_urls: string[]
}

export interface DeploymentAccount {
  accountId: string
  slackChannel: string
}

export interface EnvironmentAccount {
  accountId: string
  ownerAccountId: string
}

export interface ArtifactVersion {
  uri: string
  store: string
  path: string
  version: string
}

export interface CognitoDetails {
  /**
   * The URL where users can authenticate
   */
  authUrl: string

  /**
   * The URL for the /.well-known/jwks.json endpoint
   */
  jwksUrl: string

  /**
   * The URL for the /.well-known/openid-configuration endpoint
   */
  openIdUrl: string

  /**
   * The URI for the issuer
   */
  issuer: string
}

/**
 * Custom Resource event types
 */
export interface CustomResourceRequest {
  RequestType: "Create" | "Update" | "Delete"
  RequestId: string
  ResponseURL: string
  ResourceType: string
  LogicalResourceId: string
  StackId: string
  PhysicalResourceId?: string
  ResourceProperties: Record<string, any>
  OldResourceProperties?: Record<string, any>
}

export interface CustomResourceResponse {
  Status: "SUCCESS" | "FAILED"
  Reason?: string
  PhysicalResourceId: string
  StackId: string
  RequestId: string
  LogicalResourceId: string
  Data?: Record<string, any>
}
