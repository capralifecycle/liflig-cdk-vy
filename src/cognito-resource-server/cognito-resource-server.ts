/**
 * CDK Construct for Cognito Resource Server
 */

import { createRequire } from "node:module"
import * as path from "node:path"
import { fileURLToPath } from "node:url"
import * as cdk from "aws-cdk-lib"
import { Construct } from "constructs"
import type { ResourceServerProvider } from "../vy-cognito-provider"

const require = createRequire(import.meta.url)
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

export interface Scope {
  /**
   * The name of the scope
   */
  readonly name: string

  /**
   * A description of what this scope is for
   */
  readonly description: string
}

export interface CognitoResourceServerProps {
  /**
   * An ResourceServerProvider provided from a VyCognitoProvider
   */
  readonly resourceServerProvider: ResourceServerProvider

  /**
   * The name of the resource server
   */
  readonly name: string

  /**
   * The identifier for this resource server (usually a URL)
   * @example 'https://api.vydev.io'
   */
  readonly identifier: string

  /**
   * Custom scopes for this resource server
   * @default - No scopes
   */
  readonly scopes?: Scope[]

  /**
   * Base domain for Cognito service
   * @default 'cognito.vydev.io'
   */
  readonly cognitoBaseDomain?: string
}

/**
 * A Cognito Resource Server managed through Vy's central Cognito service
 *
 * A resource server is an integration between a user pool and an API.
 * Each resource server has custom scopes that you must activate in your app client.
 * When you configure a resource server, your app can generate access tokens with
 * OAuth scopes that authorize read and write operations to an API server.
 *
 * @example
 * ```typescript
 *  // Create a VyCognitoProvider
 * const vyCognitoProvider = new VyCognitoProvider(this, 'MyProvider', {
 *   environment: VyEnvironment.TEST,
 * });
 *
 * const resourceServer = new CognitoResourceServer(this, 'ApiResourceServer', {
 *   resourceServerProvider: vyCognitoProvider.resourceServerProvider,
 *   name: 'my-api',
 *   identifier: 'https://my-api.vydev.io',
 *   scopes: [
 *     { name: 'read', description: 'Read access to the API' },
 *     { name: 'write', description: 'Write access to the API' }
 *   ]
 * });
 * ```
 */
export class CognitoResourceServer extends Construct {
  /**
   * The identifier of the resource server
   */
  public readonly identifier: string

  /**
   * The name of the resource server
   */
  public readonly name: string

  /**
   * The underlying custom resource
   */
  public readonly resource: cdk.CustomResource

  constructor(scope: Construct, id: string, props: CognitoResourceServerProps) {
    super(scope, id)

    this.identifier = props.identifier
    this.name = props.name

    this.resource = new cdk.CustomResource(this, "Resource", {
      serviceToken: props.resourceServerProvider.serviceToken,
      properties: {
        Environment: props.resourceServerProvider.environment,
        Name: props.name,
        Identifier: props.identifier,
        Scopes: props.scopes?.map((s) => ({
          Name: s.name,
          Description: s.description,
        })),
      },
      resourceType: "Custom::VyCognitoResourceServer",
    })
  }

  /**
   * Get a reference to a scope in the format expected by app clients
   * @param scopeName The name of the scope
   * @returns The full scope identifier (e.g., 'https://api.vydev.io/read')
   */
  public scopeIdentifier(scopeName: string): string {
    return `${this.identifier}/${scopeName}`
  }
}
