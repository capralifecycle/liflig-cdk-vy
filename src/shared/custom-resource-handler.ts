/**
 * Base handler for Custom Resources
 * Handles the CloudFormation response protocol
 */

import type { CustomResourceResponse } from "./types"

export function createSuccessResponse(
  physicalResourceId: string,
  data?: Record<string, any>,
): Pick<CustomResourceResponse, "Status" | "PhysicalResourceId" | "Data"> {
  return {
    Status: "SUCCESS",
    PhysicalResourceId: physicalResourceId,
    Data: data,
  }
}

export function createFailureResponse(
  physicalResourceId: string,
  reason: string,
): Pick<CustomResourceResponse, "Status" | "PhysicalResourceId" | "Reason"> {
  return {
    Status: "FAILED",
    PhysicalResourceId: physicalResourceId,
    Reason: reason,
  }
}

/**
 * Base error handler for custom resource operations
 */
export function handleError(error: unknown): string {
  if (error instanceof Error) {
    return error.message
  }
  return String(error)
}
