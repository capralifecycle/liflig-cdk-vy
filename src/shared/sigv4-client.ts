/**
 * Shared utility for making AWS SigV4 signed HTTP requests
 */

import { Sha256 } from "@aws-crypto/sha256-js"
import { defaultProvider } from "@aws-sdk/credential-provider-node"
import { HttpRequest } from "@smithy/protocol-http"
import { SignatureV4 } from "@smithy/signature-v4"

export interface SignedRequestOptions {
  method: string
  hostname: string
  path: string
  body?: string
  headers?: Record<string, string>
  region?: string
}

export interface SignedResponse {
  statusCode: number
  body: string
  headers: Record<string, string>
}

export async function signedRequest(
  options: SignedRequestOptions,
): Promise<SignedResponse> {
  const { method, hostname, path, body, headers = {}, region } = options

  const request = new HttpRequest({
    method,
    protocol: "https:",
    hostname,
    path,
    headers: {
      "Content-Type": "application/json",
      Host: hostname,
      ...headers,
    },
    body,
  })

  const credentialsProvider = defaultProvider()
  const credentials = await credentialsProvider()

  const signer = new SignatureV4({
    credentials,
    region: region ?? "eu-west-1",
    service: "execute-api",
    sha256: Sha256,
  })

  const signedRequest = await signer.sign(request)

  const url = `https://${signedRequest.hostname}${signedRequest.path}`
  const response = await fetch(url, {
    method: signedRequest.method,
    headers: signedRequest.headers as Record<string, string>,
    body: signedRequest.body,
  })

  const responseBody = await response.text()

  return {
    statusCode: response.status,
    body: responseBody,
    headers: Object.fromEntries(response.headers.entries()),
  }
}

/**
 * Helper to create base URL from environment
 */
export function createUrlFromEnvironment(
  baseUrl: string,
  urlPrefix: string,
  environment: string,
): string {
  if (environment === "prod") {
    return `${urlPrefix}.${baseUrl}`
  }
  return `${urlPrefix}.${environment}.${baseUrl}`
}
