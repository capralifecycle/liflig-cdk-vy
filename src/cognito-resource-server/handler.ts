/**
 * Lambda handler for CognitoResourceServer custom resource
 */

import {
  createFailureResponse,
  createSuccessResponse,
  handleError,
} from "../shared/custom-resource-handler"
import { createUrlFromEnvironment, signedRequest } from "../shared/sigv4-client"
import type {
  CustomResourceRequest,
  CustomResourceResponse,
  ResourceServer,
  ResourceServerUpdateRequest,
} from "../shared/types"

const COGNITO_BASE_DOMAIN =
  process.env.COGNITO_BASE_DOMAIN || "cognito.vydev.io"

interface ResourceServerProperties {
  Environment: string
  Name: string
  Identifier: string
  Scopes?: Array<{ Name: string; Description: string }>
}

async function createResourceServer(
  baseUrl: string,
  server: ResourceServer,
): Promise<ResourceServer> {
  const response = await signedRequest({
    method: "POST",
    hostname: baseUrl,
    path: "/resource-servers",
    body: JSON.stringify(server),
  })

  if (response.statusCode !== 201) {
    throw new Error(
      `Could not create resource: ${response.statusCode} - ${response.body}`,
    )
  }

  return JSON.parse(response.body)
}

async function readResourceServer(
  baseUrl: string,
  identifier: string,
): Promise<ResourceServer> {
  const encodedIdentifier = encodeURIComponent(identifier)
  const response = await signedRequest({
    method: "GET",
    hostname: baseUrl,
    path: `/resource-servers/${encodedIdentifier}`,
  })

  if (response.statusCode !== 200) {
    throw new Error(
      `Could not read resource: ${response.statusCode} - ${response.body}`,
    )
  }

  return JSON.parse(response.body)
}

async function updateResourceServer(
  baseUrl: string,
  update: ResourceServerUpdateRequest,
): Promise<void> {
  const encodedIdentifier = encodeURIComponent(update.identifier)
  const response = await signedRequest({
    method: "PUT",
    hostname: baseUrl,
    path: `/resource-servers/${encodedIdentifier}`,
    body: JSON.stringify(update),
  })

  if (response.statusCode !== 200) {
    throw new Error(
      `Could not update resource: ${response.statusCode} - ${response.body}`,
    )
  }
}

async function deleteResourceServer(
  baseUrl: string,
  identifier: string,
): Promise<void> {
  const encodedIdentifier = encodeURIComponent(identifier)
  const response = await signedRequest({
    method: "DELETE",
    hostname: baseUrl,
    path: `/resource-servers/${encodedIdentifier}`,
  })

  if (response.statusCode !== 200) {
    throw new Error(
      `Could not delete resource: ${response.statusCode} - ${response.body}`,
    )
  }
}

export async function handler(
  event: CustomResourceRequest,
): Promise<CustomResourceResponse> {
  const props = event.ResourceProperties as ResourceServerProperties
  const baseUrl = createUrlFromEnvironment(
    COGNITO_BASE_DOMAIN,
    "delegated",
    props.Environment,
  )

  try {
    switch (event.RequestType) {
      case "Create": {
        const server: ResourceServer = {
          identifier: props.Identifier,
          name: props.Name,
          scopes: props.Scopes?.map((s) => ({
            name: s.Name,
            description: s.Description,
          })),
        }

        const created = await createResourceServer(baseUrl, server)

        return createSuccessResponse(
          event.PhysicalResourceId ?? created.identifier,
          {
            Identifier: created.identifier,
            Name: created.name,
            Scopes: created.scopes,
          },
        ) as CustomResourceResponse
      }

      case "Update": {
        const update: ResourceServerUpdateRequest = {
          identifier: props.Identifier,
          name: props.Name,
          scopes: props.Scopes?.map((s) => ({
            name: s.Name,
            description: s.Description,
          })),
        }

        await updateResourceServer(baseUrl, update)
        const updated = await readResourceServer(baseUrl, props.Identifier)

        return createSuccessResponse(
          event.PhysicalResourceId ?? updated.identifier,
          {
            Identifier: updated.identifier,
            Name: updated.name,
            Scopes: updated.scopes,
          },
        ) as CustomResourceResponse
      }

      case "Delete": {
        const identifier = event.PhysicalResourceId || props.Identifier
        await deleteResourceServer(baseUrl, identifier)

        return createSuccessResponse(identifier, {}) as CustomResourceResponse
      }
    }
  } catch (error) {
    console.error("Error:", error)
    return createFailureResponse(
      event.PhysicalResourceId || props.Identifier || "unknown",
      handleError(error),
    ) as CustomResourceResponse
  }
}
