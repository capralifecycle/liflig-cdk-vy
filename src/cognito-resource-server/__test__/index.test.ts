import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { VyEnvironment } from "../../shared/types"
import type { ResourceServerProvider } from "../../vy-cognito-provider"
import { CognitoResourceServer } from "../cognito-resource-server"

test("cognito resource server", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const mockResourceServerProvider: ResourceServerProvider = {
    environment: VyEnvironment.TEST,
    serviceToken: "serviceToken",
  }

  new CognitoResourceServer(stack, "CognitoResourceServer", {
    resourceServerProvider: mockResourceServerProvider,
    name: "testName",
    identifier: "testIdentifier",
    scopes: [
      {
        name: "testScopeName",
        description: "testScopeDescription",
      },
    ],
  })

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})
