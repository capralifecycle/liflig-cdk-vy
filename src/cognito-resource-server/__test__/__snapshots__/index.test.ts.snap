// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`cognito resource server 1`] = `
Object {
  "Resources": Object {
    "CognitoResourceServer0B86DEED": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Environment": "test",
        "Identifier": "testIdentifier",
        "Name": "testName",
        "Scopes": Array [
          Object {
            "Description": "testScopeDescription",
            "Name": "testScopeName",
          },
        ],
        "ServiceToken": "serviceToken",
      },
      "Type": "Custom::VyCognitoResourceServer",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
